### PR TITLE
[core] Fix clang-tidy modernize-* warnings 

### PR DESCRIPTION
**Pull request by tmpsantos**
Also add it to warnings as errors list so they don't come back.

### CODE REVIEW

File: src/mbgl/map/map.cpp [hunk 1]
```diff
@@ -187,6 +187,7 @@ CameraOptions Map::cameraForLatLngBounds(const LatLngBounds& bounds,
 
 CameraOptions cameraForLatLngs(const std::vector<LatLng>& latLngs, const Transform& transform, const EdgeInsets& padding) {
     if (latLngs.empty()) {
+        // NOLINTNEXTLINE(modernize-return-braced-init-list)
```
Comment (tmpsantos) [inline src/mbgl/map/map.cpp]: g++4.9 doesn't like this one as `{}`

File: src/mbgl/style/expression/compound_expression.cpp [hunk 2]
```diff
@@ -267,12 +265,10 @@ optional<std::string> featurePropertyAsString(const EvaluationContext& params, c
...
-    return id.match(
-        [](double value) { return value; },
-        [](uint64_t value) { return optional<double>(static_cast<double>(value)); },
-        [](int64_t value) { return optional<double>(static_cast<double>(value)); },
-        [](auto) { return optional<double>(); }
-    );
+    return id.match([](double value) { return value; },
+                    [](uint64_t value) -> optional<double> { return {static_cast<double>(value)}; },
+                    [](int64_t value) -> optional<double> { return {static_cast<double>(value)}; },
+                    [](auto) -> optional<double> { return {}; });
```
Comment (alexshalamov) [inline src/mbgl/style/expression/compound_expression.cpp]: const auto&

File: src/mbgl/style/expression/compound_expression.cpp [hunk 3]
```diff
@@ -246,12 +246,10 @@ optional<double> featurePropertyAsDouble(const EvaluationContext& params, const
...
-    return property->match(
-        [](double value) { return value; },
-        [](uint64_t value) { return optional<double>(static_cast<double>(value)); },
-        [](int64_t value) { return optional<double>(static_cast<double>(value)); },
-        [](auto) { return optional<double>(); }
-    );
+    return property->match([](double value) { return value; },
+                           [](uint64_t value) -> optional<double> { return {static_cast<double>(value)}; },
+                           [](int64_t value) -> optional<double> { return {static_cast<double>(value)}; },
+                           [](auto) -> optional<double> { return {}; });
```
Comment (alexshalamov) [inline src/mbgl/style/expression/compound_expression.cpp]: const auto&

Comment (alexshalamov) [reply src/mbgl/style/expression/compound_expression.cpp]: there are other places where we make unnecessary copies :(
maybe replace all of them in this PR?
```
[](auto) -> [](const auto&)
[](std::string) -> [](const std::string&)
```

Comment (tmpsantos) [reply src/mbgl/style/expression/compound_expression.cpp]: Found two more places like this

File: src/mbgl/map/map.cpp (hunk 1 above)
Comment (tmpsantos) [reply src/mbgl/map/map.cpp]: removed gcc4.9

**Pull request merged**